### ISSUE TITLE
How can I install `k2` with version `1.9.dev20211101`.

### ISSUE THREAD

**Issue by TeaPoly** [k2-fsa/icefall#161]
https://github.com/k2-fsa/icefall/blob/14c93add507982306f5a478cd144e0e32e0f970d/.github/workflows/run-pretrained-transducer-stateless.yml#L36

Thanks for the great work before I ask question.
Here is my problem. I want use `icefall` for stateless transducer experiment, it requires `torchaudio 0.10.0`, it means we need `PyTorch 1.10.0`. 
But I can't find any specific wheel package version from `conda` or `pip` . So I guess we need install from code, but I don't know how to checkout `k2 version == 1.9.dev20211101`. 

Hope someone can help me. Thanks.

**Comment by csukuangfj** [comment]
Could you first refer to https://k2-fsa.github.io/k2/installation/from_source.html
or
https://k2-fsa.github.io/k2/installation/for_developers.html
to install k2 from source?

> So I guess we need install from code, but I don't know how to checkout k2 version == 1.9.dev20211101.

Just using the master of k2 is fine.

---

Will remove the constraint of using torchaudio 0.10.0 soon so that you can also use PyTorch versions, e.g., 1.7.1

**Comment by csukuangfj** [comment]
https://github.com/k2-fsa/icefall/blob/14c93add507982306f5a478cd144e0e32e0f970d/.github/workflows/run-pretrained-transducer-stateless.yml#L36

is for CI to run inference, which does not need to compute the transducer loss. Hence, it does not depend on torchaudio 0.10.0.  Currently, only the training code needs torchaudio 0.10.0.

**Comment by TeaPoly** [comment]
> Could you first refer to https://k2-fsa.github.io/k2/installation/from_source.html or https://k2-fsa.github.io/k2/installation/for_developers.html to install k2 from source?
> 
> > So I guess we need install from code, but I don't know how to checkout k2 version == 1.9.dev20211101.
> 
> Just using the master of k2 is fine.
> 
> Will remove the constraint of using torchaudio 0.10.0 soon so that you can also use other PyTorch versions, e.g., 1.7.1

Thanks for you reply. I have tried to compile k2, but it failed with some exception like this: 

```shell
/home/huanglk/anaconda3/envs/k2/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/nn/cloneable.h: In instantiation of ‘std::shared_ptr<torch::nn::Module> torch::nn::Cloneable<Derived>::clone(const c10::optional<c10::Device>&) const [with Derived = torch::nn::CrossMapLRN2dImpl]’:
/tmp/tmpxft_000077ff_00000000-6_pytorch_context.compute_86.cudafe1.stub.c:16:27:   required from here
/home/huanglk/anaconda3/envs/k2/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/nn/cloneable.h:57:59: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
/home/huanglk/anaconda3/envs/k2/lib/python3.8/site-packages/torch/include/torch/csrc/api/include/torch/nn/cloneable.h:69:61: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >’ to type ‘torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >&’
```

 ENV:
```
gcc version 5.5.0 20171010 (Ubuntu 5.5.0-12ubuntu1~16.04)
cmake version 3.20.5
Python 3.8.12 (default, Oct 12 2021, 13:49:34)
PyTorch 1.10.0
CUDA toolkit 11.1
cuDNN 8.3.1
```

**Comment by danpovey** [comment]
That same error is mentioned here https://github.com/NVIDIA/apex/issues/805
where it says it can be resolved by upgrading to gcc 7.
Of course, it would be more ideal if the cmake files figured out the problem (I'd have thought that torch's cmake
package or whatever would notice this, if it were an incompatibility).

**Comment by csukuangfj** [comment]
The same error that prevents GitHub actions from building k2 with torch 1.10 when using ubuntu 16.04.

As Dan suggested, using GCC 7 will fix it; k2 is also using GCC 7 to build conda packages against torch 1.10.
See
https://github.com/k2-fsa/k2/blob/master/scripts/conda/k2/meta.yaml#L26
```
    - gcc_linux-64=7
```

**Comment by TeaPoly** [comment]
Thanks for your suggestion. It resolved by upgrading to gcc 7.
